### PR TITLE
dlb: optimize auto detect logic

### PR DESCRIPTION
**Pull request by daixiang0**
Signed-off-by: Loong <loong.dai@intel.com>

The default of type uint32 in protobuf is 0, so the `else` field can not run into.

Fix the above issue and add tests to cover some cases.

DLB mock functions are in development, so only test the config logic for now.

Commit Message: 
Additional Description:
Risk Level:
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

**Comment by daixiang0** [comment]
@adisuissa @wrowe could you help review?

File: contrib/network/connection_balance/dlb/source/connection_balancer_impl.cc [hunk 1]
```diff
@@ -40,25 +40,21 @@ DlbConnectionBalanceFactory::createConnectionBalancerFromProto(
 #ifdef DLB_DISABLED
   throw EnvoyException("X86_64 architecture is required for Dlb.");
 #else
-  int device_id = 0;
+  int device_id = dlb_config.id();
```
Comment (adisuissa) [inline contrib/network/connection_balance/dlb/source/connection_balancer_impl.cc:43]: Could this be a negative number? Should the config be rejected if it is?

**Review by adisuissa — COMMENTED** [review]
LGTM, just left a question.
RE how to test this, I don't think the test should create "/dev/..." files.
Can you refactor the code so it reads it from a different location? Then the test will use the functions from https://github.com/envoyproxy/envoy/blob/main/test/test_common/environment.h to create the temp files.

File: contrib/network/connection_balance/dlb/source/connection_balancer_impl.cc (hunk 1 above)
Comment (daixiang0) [reply contrib/network/connection_balance/dlb/source/connection_balancer_impl.cc:43]: Here the id is `uint32`, which cannot be negative, refer to https://stackoverflow.com/questions/765916/is-there-ever-a-good-time-to-use-int32-instead-of-sint32-in-google-protocol-buff

**Comment by daixiang0** [comment]
@adisuissa I have refactored it and added new tests, please review and merge if possible.

File: contrib/network/connection_balance/dlb/source/connection_balancer_impl.cc [hunk 2]
```diff
@@ -37,35 +35,17 @@ DlbConnectionBalanceFactory::createConnectionBalancerFromProto(
         "please decrease the number of threads by `--concurrency`");
   }
 
+  const int& config_id = dlb_config.id();
```
Comment (adisuissa) [inline contrib/network/connection_balance/dlb/source/connection_balancer_impl.cc:38]: Sorry for not being clear earlier.
The [id](https://github.com/envoyproxy/envoy/blob/1ad89a900b9239539820d9ff5c322414efa4fd07/api/contrib/envoy/extensions/network/connection_balance/dlb/v3alpha/dlb.proto#L32) is supposed to be unsigned int. Why is it being converted to a signed int?

File: contrib/network/connection_balance/dlb/source/connection_balancer_impl.h [hunk 3]
```diff
@@ -49,6 +51,33 @@ class DlbBalancedConnectionHandlerImpl : public Envoy::Network::BalancedConnecti
   Envoy::Event::FileEventPtr dlb_event_;
 };
 
+// The dir should always be "/dev" in production.
+// For test it is a temporary directory.
+// Return Dlb device id, -1 means error.
+static int detectDlbDevice(const int& config_id, const std::string& dir) {
```
Comment (adisuissa) [inline contrib/network/connection_balance/dlb/source/connection_balancer_impl.h:57]: nit: no need to pass int by ref

File: contrib/network/connection_balance/dlb/source/connection_balancer_impl.cc (hunk 2 above)
Comment (daixiang0) [reply contrib/network/connection_balance/dlb/source/connection_balancer_impl.cc:38]: Sure, I will update it.

File: contrib/network/connection_balance/dlb/source/connection_balancer_impl.h [hunk 4]
```diff
@@ -49,6 +51,33 @@ class DlbBalancedConnectionHandlerImpl : public Envoy::Network::BalancedConnecti
   Envoy::Event::FileEventPtr dlb_event_;
 };
 
+// The dir should always be "/dev" in production.
+// For test it is a temporary directory.
+// Return Dlb device id, absl::nullopt means error.
+static absl::optional<uint> detectDlbDevice(const uint config_id, const std::string& dir) {
```
Comment (adisuissa) [inline contrib/network/connection_balance/dlb/source/connection_balancer_impl.h:57]: nit: you can use `absl::StatusOr` as it allows you to specify the error.

**Review by adisuissa — APPROVED** [review]
LGTM, left a comment for future reference.

File: contrib/network/connection_balance/dlb/source/connection_balancer_impl.h (hunk 4 above)
Comment (daixiang0) [reply contrib/network/connection_balance/dlb/source/connection_balancer_impl.h:57]: Sorry for slow response due to sickness, I will optimize it in the next PR.

**Pull request merged**